[Jun-How] iP

Jun-How opened the pull request:
# JeoBot

_Struggling to keep your life organised?_
JeoBot is the right tool for you!

JeoBot keeps you worry-free by keeping track of the things you need to do.

### Plus Points ✔️
- Text-based
- User-friendly
- Convenient
- ~~FAST~~ BREAKNECK FAST to use
- What's more, it is **FREE**!

### Activation
1. Download the latest JAR release file from [here](https://github.com/Jun-How/ip/releases/tag/A-Release).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

### Features
- [x] Managing different tasks (ToDo, Deadline, Event)
- [x] Finding tasks based on keyword or due date
- [x] Adding tags for customisation
- [x] GUI support provided
- [ ] Reminders (coming soon)

If you are a Java programmer, you may use it to practise Java too! Here's the main method:

```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```
_What are you waiting for? Download it already!_

Commit:
Add Gradle support

Jun-How committed:
Implemented Level-1 increment.Renamed bot.

Jun-How committed:
Added Storage class

Jun-How committed:
Implemented Level-2

Jun-How committed:
Added Task class (A-Classes)

Jun-How committed:
Implemented Level-3

Jun-How committed:
Merge branch 'master' of https://github.com/Jun-How/ip

Jun-How committed:
Implemented Level-4 and A-Inheritance

Jun-How committed:
Implemented A-TextUiTesting

Jun-How committed:
Implemented Level-5 and A-Exceptions

Jun-How committed:
Renamed Duke

Jun-How committed:
Implemented Level-6 and A-Collections

Jun-How committed:
Implemented A-Enums

Jun-How committed:
Update Javadocs

Jun-How committed:
Add getters

Jun-How committed:
Implement save and load

Jun-How committed:
Implement date-time parsing and formatting

Jun-How committed:
Implement due command

Calling due with a date prints the list of tasks due on the date

Jun-How committed:
Resolve merge conflicts

Jun-How committed:
Minor change to add Level-8 tag

Jun-How committed:
Refactor code to include more OOP

Jun-How committed:
Organized classes into packages

Jun-How committed:
Test build gradle

Jun-How committed:
Resolve merge conflicts for gradle

Jun-How committed:
Resolve merge conflicts

Jun-How committed:
Automate project builds using Gradle, Add some JUnit Tests

Jun-How committed:
Change class name

Jun-How committed:
Add JavaDocs

Jun-How committed:
Tweak code to comply with the module coding standard

Jun-How committed:
Implement Find

Jun-How committed:
Update JavaDoc

Jun-How committed:
Removed redundant comments

Jun-How committed:
Merge branch 'branch-A-JavaDoc'

Jun-How committed:
Resolve merge conflicts

Jun-How committed:
Merge branch 'branch-Level-9'

Jun-How committed:
Update JavaDoc

Jun-How committed:
Changed some namings

Jun-How committed:
Fix coding style violations using CheckStyle

Jun-How committed:
Add GUI support to JeoBot using JavaFX

Jun-How committed:
Fix additional CheckStyle errors

Jun-How committed:
Add JavaDoc

Jun-How committed:
Edit main class name

Diatbbin commented on src/main/java/JeoBot.java:
```diff
@@ -0,0 +1,146 @@
+    protected static final String DATE_PARSE = "yyyy-MM-dd";
+    protected Ui ui;
+    protected Storage store;
+    protected TaskList taskList;
```
Would it be better to set the access modifiers as private instead?

Diatbbin commented on src/main/java/JeoBot.java:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * Represents the list of commands recognised by the bot.
+     */
+    public enum Command {
+        BYE, LIST, MARK, UNMARK, DELETE, TODO, DEADLINE, EVENT, DUE, FIND
+    }
```
Good use of enums!!

Diatbbin commented on src/main/java/jeo/parser/Parser.java:
```diff
@@ -0,0 +1,182 @@
+            str = str.substring(4).trim();
+            hm.put("description", str);
+            if (str.isEmpty()) {
+                throw new JeoException("Please enter a task description.");
+            }
+        } else if (str.toLowerCase().startsWith("deadline")) {
+            hm.put("command", "deadline");
+            str = str.substring(8).trim();
+            if (!str.toLowerCase().contains("/by")) {
+                throw new JeoException(str.isEmpty()
+                        ? "Please enter a task description."
+                        : "Please follow the format: deadline <description> /by <yyyy-MM-dd HH:mm>");
+            }
+            String[] arr = parseSubstringTasks(str, "deadline");
+            String desc = arr[0];
+            String by = arr[1];
+            if (desc.isEmpty()) {
+                throw new JeoException("Please enter a task description.");
+            }
+            if (by.isEmpty()) {
+                throw new JeoException("Please enter a date-time in the format "
+                        + "\"yyyy-MM-dd HH:mm\" after \"/by\".");
+            }
+            hm.put("description", desc);
+            hm.put("by", by);
+        } else if (str.toLowerCase().startsWith("event")) {
+            hm.put("command", "event");
+            str = str.substring(5).trim();
+            if (!str.toLowerCase().contains("/from") || !str.toLowerCase().contains("/to")) {
+                throw new JeoException(str.isEmpty()
+                        ? "Please enter a task description."
+                        : "Please follow the format: event <description> "
+                        + "/from <yyyy-MM-dd HH:mm> /to <yyyy-MM-dd HH:mm>");
+            }
+            String[] arr = parseSubstringTasks(str, "event1");
+            String desc = arr[0];
+            str = arr[1];
+            arr = parseSubstringTasks(str, "event2");
+            String from = arr[0];
+            String to = arr[1];
```
I think u can consider replacing the if-else structure with switch case for better code readability.

Diatbbin reviewed (commented):
:) 👍 Overall, really good use of OOP. Class structures seem to be well thought out. Coding style and variable namings looks good too! I think Parser class was abit hard to read, I think u can consider using functions to remove duplicated code.

Jun-How committed:
Make dummy change

ZHTang29 commented on src/main/java/jeo/parser/Parser.java:
```diff
@@ -0,0 +1,182 @@
+                throw new JeoException("Please enter a date-time in the format "
+                        + "\"yyyy-MM-dd HH:mm\" after \"/from\".");
+            }
+            if (to.isEmpty()) {
+                throw new JeoException("Please enter a date-time in the format "
+                        + "\"yyyy-MM-dd HH:mm\" after \"/to\".");
+            }
+            hm.put("description", desc);
+            hm.put("from", from);
+            hm.put("to", to);
+        } else if (str.toLowerCase().startsWith("mark")) {
+            hm.put("command", "mark");
+            int i = parseSubStringActions(str, 4);
+            hm.put("index", Integer.toString(i - 1));
+        } else if (str.toLowerCase().startsWith("unmark")) {
+            hm.put("command", "unmark");
+            int i = parseSubStringActions(str, 6);
+            hm.put("index", Integer.toString(i - 1));
+        } else if (str.toLowerCase().startsWith("delete")) {
+            hm.put("command", "delete");
+            int i = parseSubStringActions(str, 6);
+            hm.put("index", Integer.toString(i - 1));
+        } else if (str.toLowerCase().startsWith("due")) {
+            hm.put("command", "due");
+            String by = str.substring(3).trim();
+            if (by.isEmpty()) {
+                throw new JeoException("Please enter a date in the format: \"yyyy-MM-dd\".");
+            }
+            hm.put("by", by);
+        } else if (str.toLowerCase().startsWith("find")) {
+            hm.put("command", "find");
+            String key = str.substring(4).trim();
+            if (key.isEmpty()) {
+                throw new JeoException("Please enter a keyword.");
+            }
+            hm.put("key", key);
+        } else {
+            hm.put("command", "");
+        }
+        return hm;
```
Similar to another comment on the same block of if-else code, with such a long block the readability is slightly compromised (and it is slightly harder to debug), so you may consider editing the code to cater to that (switch cases etc.).

ZHTang29 reviewed (commented):
Overall, coding quality is good and generally most portion of the code is readable and good use of OOP! Keep it up!

Jun-How committed:
Use assert to document important assumptions

When each input is being parsed and returned as a hashmap,
the keys should contain keywords required for the relevant command.

By recording these assumptions, it serves as a check to ensure that
the keys for the relevant command appear in the hashmap for execution.
This way, the command should execute and output as expected.

Jun-How committed:
Improve code quality

- Made use of SLAP to make code cleaner and more readable in MainWindow class.
- Fixed some complicated expressions in Ui class.
- Fixed deep nesting issue in Parser class.

Jun-How committed:
Merge pull request #1 from Jun-How/branch-A-Assertions

Use assert to document important assumptions

Jun-How committed:
Merge branch 'master' into branch-A-CodeQuality

Jun-How committed:
Merge pull request #2 from Jun-How/branch-A-CodeQuality

Improve code quality

Jun-How committed:
Add Extension of C-Tagging

Users can now add optional tags as many times as they want to tasks.
The tags can be added anywhere after the description of the task.
E.g. todo read book /tag study /tag fun

This is useful because it gives users the opportunity to customise
their own style of categorising various tasks.

Jun-How committed:
Dummy change to tag BCD-Extension

Jun-How committed:
Add better GUI

Jun-How committed:
Add product name and icon in title bar

Jun-How committed:
Update README.md

Serves as a user guide

Jun-How committed:
Update README.md

Jun-How committed:
Update version number

Jun-How committed:
Merge branch 'master' of https://github.com/Jun-How/ip

Jun-How committed:
Update README.md

Jun-How committed:
Update README.md

Jun-How committed:
Update README.md

Jun-How committed:
Make minor dummy change

Jun-How committed:
Removed Ui.png

Jun-How committed:
Re-add Ui.png file to docs

Jun-How committed:
Abstract out Command and Parser classes

Jun-How committed:
Add JavaDocs to new classes, methods and fix checkstyle errors

Jun-How committed:
Update README.md

Jun-How committed:
Update README.md

Jun-How committed:
Update README.md
